### ISSUE TITLE
Changes made to a npm linked source code are not reflected following gulp build being executed

### ISSUE THREAD

**Issue by N-Roberts** [liferay/liferay-js-themes-toolkit#499]
**Steps to reproduce:**

1. Install the latest version of the theme generator for 7.1 using: npm install -g generator-liferay-theme@^8.2.1
2. Download the repository https://github.com/lgdd/sign-in-modal-themelet and extract it
3. In the sign-in-modal-themelet folder run ’npm link’
4. Outside the themelet, create a new theme by running yo liferay-theme,

- name: test
- id: test
- version: 7.1
- deployment: other
- URL: default

5. Run ./node_modules/.bin/gulp extend from the root of the created theme

- kind: themelet
- Where: "Search globally installed npm modules (development purposes only)"
- select: sign-in-modal-themelet (pressing space bar will select the themelet and is necessary)

6. Run ./node_modules/.bin/gulp build
7. Look at the build/themelets/sign-in-modal-themelet/js/main.js file
8. In the themelet clone, make a change e.g.: changing the class .sign-in with .sign-in-test
9. In the theme folder run ’gulp build’ and look at the file build/themelets/sign-in-modal-themelet/js/main.js

**Results:**
 Expected Result: After gulp build is executed in step 9 the customer expects for the main.js class to reflect the changes made to the themelet source
 Actual Result: The main.js file does not reflect the changes made to the themelet source after step 9

**Errors/Logs**
Main.js after step 7:

```
AUI().ready(
    'liferay-sign-in-modal',
    function (A) {
        var signIn = A.one('.sign-in > a');

        if (signIn && signIn.getData('redirect') !== 'true') {
            signIn.plug(Liferay.SignInModal);
        }
    }
);
```
Main.js after step 9:

```
AUI().ready(
    'liferay-sign-in-modal',
    function (A) {
        var signIn = A.one('.sign-in > a');

        if (signIn && signIn.getData('redirect') !== 'true') {
            signIn.plug(Liferay.SignInModal);
        }
    }
);
```

**Comment by wincent** [comment]
Thanks for the clear repro steps @N-Roberts. I get exactly the results that you describe, but I think this is a mixture of a real bug in the toolkit (nothing to do with `npm link`) _plus_ a non-bug related to how `npm link` is supposed to work. There is a workaround.

1.  `npm link` from the sign-in-modal-themelet folder creates a symlink from the global `node_modules` folder to the themelet source. For example, on my machine, that means:

    `/Users/greghurrell/n/lib/node_modules/sign-in-modal-themelet@ -> /Users/greghurrell/code/sign-in-modal-themelet`

2.  When you run `gulp extend` in your theme, `sign-in-modal-themelet` is added to the `dependencies` field of the `package.json` file, and the themelet files are _copied_ (not linked) to `node_modules/sign-in-modal-themelet`, just as if you had run `npm install sign-in-modal-themelet`:

    ```
    drwxr-xr-x    6 greghurrell  staff    192  3 Aug 11:48 sign-in-modal-themelet/
    ```

    This is the bit that I refer to as "a real bug" above; the package.json correctly refers to the local path — eg. `"path": "/Users/greghurrell/n/lib/node_modules/sign-in-modal-themelet"` on my machine — in the `liferayTheme` settings:

    ```
    "liferayTheme": {
            "baseTheme": "styled",
            "screenshot": "",
            "templateLanguage": "ftl",
            "version": "7.1",
            "themeletDependencies": {
                    "sign-in-modal-themelet": {
                            "liferayTheme": {
                                    "screenshot": "",
                                    "themelet": true,
                                    "version": "*"
                            },
                            "name": "sign-in-modal-themelet",
                            "version": "1.0.2",
                            "path": "/Users/greghurrell/n/lib/node_modules/sign-in-modal-themelet"
                    }
            }
    },
    ```

    _but_ it refers to the themelet package by its registry name in the `dependencies` section, and the package-lock.json shows it being resolved from the remote host:

    ```
    "sign-in-modal-themelet": {
        "version": "1.0.2",
        "resolved": "https://registry.npmjs.org/sign-in-modal-themelet/-/sign-in-modal-themelet-1.0.2.tgz",
        "integrity": "sha512-exkxRnRcaxJihW/g7QEKz9XvDXafIEEyBDW265gQWYjRykHf3E/Vqnu3AO2evLLydL10t+6pJ9p+1CKqZ1eeBQ=="
    },
    ```

3.  When you update the source files and build, you don't get the updated files automatically, because your theme continues to operate with the files that were copied in step 2.

---

So, about the workaround and the non-bug I mentioned up top: [the `npm link` manual says](https://docs.npmjs.com/cli/link):

> Package linking is a two-step process.

ie. running `npm link` in a source folder without running a corresponding `npm link` in a consuming folder is only half of the process. If you run `npm link sign-in-modal-themelet` in the theme folder, then everything works as you would expect:

-   Instead of a copy of the themelet files, you get a live link to the source:

    ```
    lrwxr-xr-x    1 greghurrell  staff     86  3 Aug 11:56 sign-in-modal-themelet@ -> ../../../../../../../../../Users/greghurrell/n/lib/node_modules/sign-in-modal-themelet
    ```

-   When you run `gulp build` in the theme, your build reflects whatever edits you've made to the themelet source.

So, we have a workaround: doing `npm link` without actually pairing it with a symmetrical `npm link` in the consuming theme is a user error, so you just need to run the other half of the `npm link` pair.

---

As for the bug, it's not related to `npm link`. If we ignore the `npm link` part of your repro recipe we can still see a problem: the `npm install` we run ends up pulling down the themelet from the network instead of installing form the local "global" install. The problem lies [on this line](https://github.com/liferay/liferay-js-themes-toolkit/blob/2ac32f5e26d959923c96ad49bc6438cda0d70cbc/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L172) where we check for `__realPath__` to detect a local install, but by the time we get there `__realPath__` has been [turned into `path` instead](https://github.com/liferay/liferay-js-themes-toolkit/blob/2ac32f5e26d959923c96ad49bc6438cda0d70cbc/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L345). I believe it's been doing it this way for about 15 months now, since [this commit](https://github.com/liferay/liferay-js-themes-toolkit/commit/7f083f9f49dab3777da7574a4def31631bcd67ae) — I explained in that one why I thought it was the right change to make, but now I am having second thoughts. :grinning: I'm going to look into tweaking to see if I can fix things.

**Comment by jbalsas** [comment]
For additional context, I think this was working fine in current `9.x` series and only failing in `8.x`. At least that's what I observed when trying to repro.

**Comment by wincent** [comment]
Looking at the code a bit more, I think that this only affects themelets, which probably aren't all that well tested; shouldn't be too hard to fix.

- Themes:
  - `__filePath__` becomes `path` on [line 82](https://github.com/liferay/liferay-js-themes-toolkit/blob/2ac32f5e26d959923c96ad49bc6438cda0d70cbc/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L82).
  - But when we install on [line 90](https://github.com/liferay/liferay-js-themes-toolkit/blob/2ac32f5e26d959923c96ad49bc6438cda0d70cbc/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L90), we're still using the untransformed `__filePath__` and the right thing gets installed.
- Themelets:
  - `__filePath__` gets transformed to `path` on [line 101](https://github.com/liferay/liferay-js-themes-toolkit/blob/2ac32f5e26d959923c96ad49bc6438cda0d70cbc/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L101).
  - And when we install on [line 125](https://github.com/liferay/liferay-js-themes-toolkit/blob/2ac32f5e26d959923c96ad49bc6438cda0d70cbc/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L125), we fail to use untransformed data (`__filePath__`) and so install the wrong thing.

Compared with master:

- We transform `__filePath__` to `path` on [line 90](https://github.com/liferay/liferay-js-themes-toolkit/blob/cefa75c9c64970d3e85530b1b2f1e33adcc49a64/packages/liferay-theme-tasks/theme/prompts/extend_prompt.js#L90-L94).
- But we use the untransformed data when installing on [line 111](https://github.com/liferay/liferay-js-themes-toolkit/blob/cefa75c9c64970d3e85530b1b2f1e33adcc49a64/packages/liferay-theme-tasks/theme/prompts/extend_prompt.js#L111).

And on 9.x, it's the same:

- Transform on [line 96](https://github.com/liferay/liferay-js-themes-toolkit/blob/dd3234ef391caba559f6b8578a514db56e07feae/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L96).
- Install using `__filepath__` on [line 117](https://github.com/liferay/liferay-js-themes-toolkit/blob/dd3234ef391caba559f6b8578a514db56e07feae/packages/liferay-theme-tasks/lib/prompts/extend_prompt.js#L117).

**Comment by wincent** [comment]
Fix is in the 8.2.2 release: https://github.com/liferay/liferay-js-themes-toolkit/releases/tag/v8.2.2
